misc/python/materialize/scalability/operation.py

```diff
@@ -0,0 +1,30 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from typing import Optional
+
+from psycopg import Cursor, ProgrammingError
+
+
+class Operation:
+    def __init__(self) -> None:
+        self._query: Optional[str] = None
+
+    def execute(self, cursor: Cursor) -> None:
+        if self._query is None:
+            self._query = self.query()
```

Comment:
I am storing it in a variable in case it is expensive to generate it -- some Operations use a format string to generate the query and I was wondering how slow that would be. We call `self.query()` up to 500 times per second.

Reply:
I have removed this complication.